binder/src/main/java/io/grpc/binder/internal/BinderTransportSecurity.java

```diff
@@ -178,9 +181,12 @@ ListenableFuture<Status> checkAuthorization(MethodDescriptor<?, ?> method) {
       boolean useCache = method.isSampledToLocalTracing();
       if (useCache) {
         @Nullable ListenableFuture<Status> authorization = serviceAuthorization.get(serviceName);
-        if (authorization != null) {
+        if (authorization != null && !isFailedOrCancelled(authorization)) {
+          // Authorization check exists and is a pending or successful future (even if for a failed
+          // authorization).
           return authorization;
         }
+        serviceAuthorization.remove(serviceName);
```

Comment:
Won't this also try to remove even when authorization is null?

Reply:
Obsolete (though removing a non-existing key should be harmless).